src/compiler/llvm-to-backend/CMakeLists.txt

```diff
@@ -118,19 +118,20 @@ if(WITH_SSCP_COMPILER)
   set(LLVMSPIRV_PATH ${LLVMSPIRV_INSTALLDIR}/bin/llvm-spirv)
   set(LLVMSPIRV_RELATIVE_PATH ${LLVMSPIRV_RELATIVE_INSTALLDIR}/bin/llvm-spirv)
 
-#  ExternalProject_Add(LLVMSpirvTranslator
- #   GIT_REPOSITORY https://github.com/KhronosGroup/SPIRV-LLVM-Translator
- #   GIT_TAG origin/${LLVMSPIRV_BRANCH}
- #   GIT_SHALLOW ON
- #   BUILD_ALWAYS OFF
- #   INSTALL_DIR ${LLVMSPIRV_INSTALLDIR}
- #   BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/ext/llvm-spirv
- #   CMAKE_CACHE_ARGS
- #     -DLLVM_SPIRV_BUILD_EXTERNAL:BOOL=YES
- #     -DLLVM_DIR:PATH=${LLVM_DIR}
-#		  -DCMAKE_INSTALL_PREFIX:PATH=<INSTALL_DIR>
-#    BUILD_COMMAND ${CMAKE_COMMAND} --build <BINARY_DIR> --config Release --target install
-#    )
+  ExternalProject_Add(LLVMSpirvTranslator
```

Comment:
Hi, it should not write to `/usr/local` if you set your `CMAKE_INSTALL_PREFIX` to a user-writable directory. If you want to install system-wide, you will need root privileges anyway.

The patch is necessary because llvm-spirv follows the SPIR convention where address space 0, the default address space, does not refer to the generic address space (i.e. "may point to any address space"). Instead, it assumes that address space 0 refers to the private addess space.

This is different from all other device backends in LLVM like ptx or amdgcn, and is a huge complication because the IR that clang gives us assumes address space 0 everywhere by default. This makes sense because C++ semantics are that every pointer is effectively a pointer in generic address space.
Additionally, while in a multi-pass compiler the device compiler invocation can be configured with different assumptions as the default address space, this is not so easily possible in a single-pass design as in our compiler. Here, host and device code are generated from the same compiler invocation and thus also share some features such as the default address space configuration.

So, we would need to completely rewrite the LLVM IR to exchange every occurence of address space 0 with llvm-spirv's convention for the generic address space. While we already do some modifications to address spaces to account for device/backend specifics such as the local memory address space, these changes are usually fairly isolated. On the other hand, address space 0 is used everywhere throughout the code, so rewriting all that IR requires a massive effort, and there are no preexisting LLVM passes to handle that use case to my knowledge. So it turned out that patching llvm-spirv is much, much simper, as it boils down to a 1-line change there.

Note that even if you get your system llvm-spirv to work, it wouldn't be found by our llvm-to-spirv JIT infrastructure which expects llvm-spirv to live in the directory that we have assigned it.

Reply:
Thanks for the details. Should that change be going upstream rather than be a permanent fork?

Reply:
Well, technically it's not a fork, but a patch that is broadly compatible with all llvm-spirv versions ;) But of course it would be more convenient if it were upstreamed at  some point.

See also https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1699

While there's no hard objection from upstream SPIRV-LLVM-Translator, it's also not really clear how this is supposed to work since presumably the old convention would still need to work (otherwise, all other stacks apart from us  depending on llvm-spirv will break), so there needs to be some kind of toggle (macro? command line flag?) and I think it's unclear how this would look like and if we would ti it to some other address space convention like PTX or just be an independent thing..